### PR TITLE
fix coupling of ptreltype and is_arrayorvararg_ty

### PR DESCRIPTION
**Pull request by wsmoses**

### CODE REVIEW

**Comment by wsmoses** [comment]
Fixes https://github.com/EnzymeAD/Enzyme.jl/issues/1152

File: src/compiler.jl [hunk 1]
```diff
@@ -339,10 +331,23 @@ end
...
+@inline ptreltype(::Type{Array{T,N}}) where {T,N} = T
+@inline ptreltype(::Type{Array{T, N} where N}) where {T} = T
+@inline ptreltype(::Type{Complex{T}}) where T = T
+@inline ptreltype(::Type{Tuple{Vararg{T}}}) where T = T
+
 @inline is_arrayorvararg_ty(::Type) = false
 @inline is_arrayorvararg_ty(::Type{Array{T,N}}) where {T,N} = true
 @inline is_arrayorvararg_ty(::Type{Array{T, N} where N}) where {T} = true
 @inline is_arrayorvararg_ty(::Type{Tuple{Vararg{T2}}}) where T2 = true
+@inline is_arrayorvararg_ty(::Type{Ptr{T}}) where T = T
```
Comment (vchuravy) [inline src/compiler.jl:347]: This needs to be a boolean

**Comment by matinraayai** [comment]
```
ERROR: LoadError: MethodError: no method matching ptreltype(::Type{Ptr})

Closest candidates are:
  ptreltype(::Type{Ptr{T}}) where T
   @ Enzyme ~/.julia/packages/Enzyme/0P1Ss/src/compiler.jl:334
  ptreltype(::Type{Core.LLVMPtr{T, N}}) where {T, N}
   @ Enzyme ~/.julia/packages/Enzyme/0P1Ss/src/compiler.jl:335
  ptreltype(::Type{Core.LLVMPtr{T}}) where T
   @ Enzyme ~/.julia/packages/Enzyme/0P1Ss/src/compiler.jl:336
  ...

Stacktrace:
  [1] macro expansion
    @ ./show.jl:1128 [inlined]
  [2] active_reg_inner
    @ ~/.julia/packages/Enzyme/0P1Ss/src/compiler.jl:376 [inlined]
  [3] active_reg_inner (repeats 2 times)
    @ ~/.julia/packages/Enzyme/0P1Ss/src/compiler.jl:354 [inlined]
  [4] guaranteed_const_nongen(#unused#::Type{Ptr}, world::UInt64)
    @ Enzyme.Compiler ~/.julia/packages/Enzyme/0P1Ss/src/compiler.jl:503
  [5] julia_activity_rule(f::LLVM.Function)
    @ Enzyme.Compiler ~/.julia/packages/Enzyme/0P1Ss/src/rules/activityrules.jl:70
  [6] codegen(output::Symbol, job::GPUCompiler.CompilerJob{Enzyme.Compiler.EnzymeTarget, Enzyme.Compiler.EnzymeCompilerParams}; libraries::Bool, deferred_codegen::Bool, optimize::Bool, toplevel::Bool, strip::Bool, validate::Bool, only_entry::Bool, parent_job::Nothing)
    @ Enzyme.Compiler ~/.julia/packages/Enzyme/0P1Ss/src/compiler.jl:4383
  [7] codegen
    @ ~/.julia/packages/Enzyme/0P1Ss/src/compiler.jl:4205 [inlined]
  [8] _thunk(job::GPUCompiler.CompilerJob{Enzyme.Compiler.EnzymeTarget, Enzyme.Compiler.EnzymeCompilerParams}, postopt::Bool)
    @ Enzyme.Compiler ~/.julia/packages/Enzyme/0P1Ss/src/compiler.jl:5148
  [9] _thunk
    @ ~/.julia/packages/Enzyme/0P1Ss/src/compiler.jl:5148 [inlined]
 [10] cached_compilation
    @ ~/.julia/packages/Enzyme/0P1Ss/src/compiler.jl:5182 [inlined]
 [11] (::Enzyme.Compiler.var"#488#489"{DataType, DataType, DataType, Enzyme.API.CDerivativeMode, NTuple{4, Bool}, Int64, Bool, Bool, UInt64, DataType})(ctx::LLVM.Context)
    @ Enzyme.Compiler ~/.julia/packages/Enzyme/0P1Ss/src/compiler.jl:5248
 [12] JuliaContext(f::Enzyme.Compiler.var"#488#489"{DataType, DataType, DataType, Enzyme.API.CDerivativeMode, NTuple{4, Bool}, Int64, Bool, Bool, UInt64, DataType})
    @ GPUCompiler ~/.julia/packages/GPUCompiler/U36Ed/src/driver.jl:47
 [13] #s1043#487
    @ ~/.julia/packages/Enzyme/0P1Ss/src/compiler.jl:5200 [inlined]
 [14] var"#s1043#487"(FA::Any, A::Any, TT::Any, Mode::Any, ModifiedBetween::Any, width::Any, ReturnPrimal::Any, ShadowInit::Any, World::Any, ABI::Any, ::Any, #unused#::Type, #unused#::Type, #unused#::Type, tt::Any, #unused#::Type, #unused#::Type, #unused#::Type, #unused#::Type, #unused#::Type, #unused#::Any)
    @ Enzyme.Compiler ./none:0
 [15] (::Core.GeneratedFunctionStub)(::Any, ::Vararg{Any})
    @ Core ./boot.jl:602
 [16] autodiff(::ReverseMode{false, FFIABI}, ::Const{var"#forward_solver#3"}, ::Type{Const{Nothing}}, ::Duplicated{Array{Float64, 3}}, ::Vararg{Any})
    @ Enzyme ~/.julia/packages/Enzyme/0P1Ss/src/Enzyme.jl:209
 [17] autodiff(::ReverseMode{false, FFIABI}, ::Const{var"#forward_solver#3"}, ::Duplicated{Array{Float64, 3}}, ::Duplicated{Matrix{Float64}}, ::Vararg{Duplicated{Matrix{Float64}}})
    @ Enzyme ~/.julia/packages/Enzyme/0P1Ss/src/Enzyme.jl:238
 [18] autodiff
    @ ~/.julia/packages/Enzyme/0P1Ss/src/Enzyme.jl:224 [inlined]
 [19] main()
    @ Main /scratch/raayaiardakani.m/Redbird.jl/example/redbird_enzyme.jl:78
 [20] top-level scope
    @ /scratch/raayaiardakani.m/Redbird.jl/example/redbird_enzyme.jl:85
 [21] include(fname::String)
    @ Base.MainInclude ./client.jl:478
 [22] top-level scope
    @ REPL[1]:1
in expression starting at /scratch/raayaiardakani.m/Redbird.jl/example/redbird_enzyme.jl:85
```
After the fix I get this error. I think something in my code is returning ```Ptr{Nothing}``` and I'm not sure why

File: src/compiler.jl [hunk 2]
```diff
@@ -362,11 +367,12 @@ end
         return ActiveState
     end
 
-    if T <: Ptr || T <: Core.LLVMPtr || T <: Base.RefValue || is_arrayorvararg_ty(T)
+    if T <: Ptr || T <: Core.LLVMPtr || T <: Base.RefValue || T <: Array
```
Comment (wsmoses) [inline src/compiler.jl:370]:
```suggestion
    if T <: Ptr || T <: Core.LLVMPtr || T <: Base.RefValue || T <: Array || is_arrayorvararg_ty(T)
```

**Pull request merged**